### DISCUSSION TITLE
Can I use Feature Adversaries - PyTorch with GPU?

### DISCUSSION THREAD

**Discussion by matanle51** [Trusted-AI/adversarial-robustness-toolbox#2003] [Q&A] [answered]
Hi, I defined a pytorch classfier:

criterion = torch.nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.01)
classifier = PyTorchClassifier(
                            model=model,
                            clip_values=(0, 1),
                            loss=criterion,
                            optimizer=optimizer,
                            input_shape=(3, 32, 32),
                            nb_classes=10,
                            device_type='gpu')

and a  Feature Adversaries attack in the following way:

attack = FeatureAdversariesPyTorch(
            classifier,
            layer=-1,
            delta=8/255,
            optimizer=None,
            step_size=1/255,
            max_iter=500,
        )

I pass it a source and a guide:

x_test_adv = attack.generate(data, guide)

But when I try to run it, I'm getting the following error:

`13, in FeatureAdversariesPyTorch.generate(self, x, y, **kwargs)
    211     guide_batch = torch.tensor(y[begin:end])
    212     x_adversarial[begin:end] = self._generate_batch(source_batch, guide_batch).numpy()
--> 213 return np.array(x_adversarial, dtype=x.dtype)

TypeError: Cannot interpret 'torch.float32' as a data type`

It seems that it's looking for numpy data type, but I pass it tensors since I work on GPU.
Any idea on how to solve this issue?

**Comment by beat-buesser** [comment, marked as answer]
Hi @matanle51 Yes, you should be able to use Feature Adversaries with GPUs. What is the type of your variable `data` and `guide`?

**Reply by matanle51** [reply]
Hi @beat-buesser,

Basically it's two batches from the standard CIFAR-10 dataset.
Type of batch: torch.tensor with dtype=torch.float32
Shape of batch: [128, 3, 32, 32]

Both are on the gpu, as well as the model.

**Reply by beat-buesser** [reply]
Hi @matanle51,
You have to provide them as Numpy arrays to the method `generate`. The attack will convert them to tensors and transfer them to the GPU internally.

**Reply by matanle51** [reply]
Hi @beat-buesser ,
If I try to pass the data as numpy arrays, I get the following message:
`RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!`

This is since the model is on GPU and the arrays are on cpu.

**Reply by matanle51** [reply]
@beat-buesser 
BTW -same error when trying to run your notebook example:
https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/a0411c54f7e0471696f6baeb2c73dea73beedb08/notebooks/attack_feature_adversaries_pytorch.ipynb

**Reply by beat-buesser** [reply]
Hi @matanle51 I think you have discovered a bug!

These lines in FeatureAdversariesPyTorch are missing a transfer to the device:
https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/a0411c54f7e0471696f6baeb2c73dea73beedb08/art/attacks/evasion/feature_adversaries/feature_adversaries_pytorch.py#L210-L211

The solution is to add `.to(self.estimator.device)` at the end of each line. Can you try if it solves the issue for you too?

 I'll open an issue for this bug and schedule it for the next patch release ART 1.13.1.

**Reply by matanle51** [reply]
@beat-buesser Thank you.

Yes, I changed lines 210-211, and it solved the problem.

`source_batch = torch.tensor(x[begin:end]).to(self.estimator.device)`
`guide_batch = torch.tensor(y[begin:end]).to(self.estimator.device)`

**Reply by beat-buesser** [reply]
Hi @matanle51 Thank you!
